devito/finite_differences/derivative.py

```diff
@@ -289,6 +297,8 @@ def _eval_fd(self, expr):
         else:
             res = generic_derivative(expr, *self.dims, self.fd_order, self.deriv_order,
                                      matvec=self.transpose, x0=self.x0)
+        if expr.is_Mul:
```

Comment:
same here, plus increases the number of evaluations... I understand it's for a fix, but isn't there another way?

Reply:
no

Reply:
so if it's a Mul you evaluate later, otherwise you evaluate before... this is IMHO an impenetrable mess without proper explanation